### ISSUE TITLE
With -d:danger, non power-of-2 table can hang

### ISSUE THREAD

**Issue by hashbackup** [nim-lang/Nim#14895]
### Example
```
ms:nim jim$ cat badtable.nim
import tables

const
  ents = 20

var
  map = newTable[int, int](ents)

for i in 0..<ents:
  echo i
  map[i] = i

ms:nim jim$ nim c -d:release badtable.nim
Hint: 22720 LOC; 0.743 sec; 24.516MiB peakmem; Release build; proj: /Users/jim/nim/badtable.nim; out: /Users/jim/nim/badtable [SuccessX]

ms:nim jim$ ./badtable
fatal.nim(49)            sysFatal
Error: unhandled exception: tableimpl.nim(127, 12) `isPowerOfTwo(initialSize)`  [AssertionError]

ms:nim jim$ nim c -d:danger badtable.nim
Hint: 22720 LOC; 0.686 sec; 24.617MiB peakmem; Dangerous Release build; proj: /Users/jim/nim
/badtable.nim; out: /Users/jim/nim/badtable [SuccessX]

ms:nim jim$ ./badtable
0
1
2
3
4
^CSIGINT: Interrupted by Ctrl-C.
ms:nim jim$
```

### Current Output
Hangs

### Expected Output
Work or fail, but don't hang

### Possible Solution
rightSize(initialSize) should always be used instead of using raw initialSize.  Users only know how many entries they want, not how big a table needs to be to accommodate that, especially since this is an implementation detail.  The next implementation of tables could require a prime number table size; then programs calling initTable with nextPowerOfTwo would fail.

### Additional Information
There is a compatibility issue because users are already calling initTable with a power-of-2 size.  If that came from rightSize, it is correct and should be used directly, because if another call is made to rightSize, it will double the size.  However, if it is an integer or from nextPowerOfTwo, the size is likely incorrect because it didn't take into account the table load factor like rightSize does.

To handle this, rightSize could return a distinct int type so that initTable would know whether rightSize needs to be called on initialSize.  Something like this:
```
ms:nim jim$ cat rightsize.nim
import math

type RS = distinct int

proc rightSize*(count: Natural): RS {.inline.} =
  ## Return the value of ``initialSize`` to support ``count`` items.
  ##
  ## If more items are expected to be added, simply add that
  ## expected extra amount to the parameter before calling this.
  #
  # Make sure to synchronize with `mustRehash`
  result = RS(nextPowerOfTwo(count * 3 div 2 + 4))

proc init(initialSize: int | RS) =
  var tableSize: int

  if type(initialSize) is RS:
    tableSize = int(initialSize)
  else:
    tableSize = int(rightSize(Natural(initialSize)))
  echo "tableSize is ", tableSize

init(16)
init(nextPowerOfTwo(16))
init(rightSize(16))

ms:nim jim$ ./rightsize
tableSize is 32
tableSize is 32
tableSize is 32
ms:nim jim$

ms:nim jim$ nim -v
Nim Compiler Version 1.2.1 [MacOSX: amd64]
Compiled at 2020-05-19
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: 9b510643faeb7eb072af83fa14e7d1a86116c65f
active boot switches: -d:release
```

Just as a sidenote, I had to add Natural(initialSize) to the rightSize call, which I don't understand.  Without it, this compile error occurs:
```
ms:nim jim$ nim c rightsize
/Users/jim/nim/rightsize.nim(25, 5) template/generic instantiation of `init` from here
/Users/jim/nim/rightsize.nim(20, 30) Error: type mismatch: got <RS>
but expected one of: 
proc rightSize(count: Natural): RS
  first type mismatch at position: 1
  required type for count: Natural
  but expression 'initialSize' is of type: RS

expression: rightSize(initialSize)
```
I don't understand why Nim thinks initialSize is of type RS here.

**Comment by jibal** [comment]
> I don't understand why Nim thinks initialSize is of type RS here.

Line 25 is `init(rightSize(16))` which results in `rightSize(rightSize(16))` on line 20, so the error message is correct ... the `if` statement has no effect on what the compiler sees, and what it sees has the wrong type. Change it to `when`--which is correct because `type(...)` is a compile-time constant--and then it works without `Natural`. Of course this means that the type of the value being passed to `init` must be known at compile time, but that's always necessary with type classes, which are compile time constructs that are just a convenience (you can get the same effect by overloading `init` with different parameter types).

BTW, `type` is discouraged; use `typeof`. (I'm surprised this hasn't been deprecated.)

**Comment by hashbackup** [comment]
Thanks for the tip, this worked great:
```
  tablesize = int(when typeof(initialSize) is RS: initialSize else: rightSize(initialSize))
```
I also missed the significance of the previous compiler note showing which line generated the bad call (I'm a neoNim).

You reminded me, I'd suggest that rightSize be deprecated as a part of this issue.

**Comment by timotheecour** [comment]
* for the 1st part (hanging), replacing `assert isPowerOfTwo(size)` by `doAssert isPowerOfTwo(size)` (+ in other relevant places in code) would fix it; the overhead will be negligible matter in the context where this is used.

* for the 2nd part, I agree but there's a difficulty:
if you replace
`assert isPowerOfTwo(size)` by `var size = rightSize(size)`
then it'd be a performance breaking change when user passes a power of 2.

if you replace `assert isPowerOfTwo(size)` by `var size = if isPowerOfTwo(size): size else: rightSize(size)` then you get a weird behavior when passing size=31 results in 64 whereas size=32 results in 32, which is not good.

so to fix the 2nd part, you'd need a separate API to distinguish whether you're specifying the size of the table or the expected number of elements; eg:
```nim
proc initTable*[A, B](initialSize = defaultInitialSize): Table[A, B] =
=>
proc initTable*[A, B](initialSize = defaultInitialSize, expectedNum = -1): Table[A, B] =
  if expectedNum == -1: # use initialSize
  else: use rightSize(expectedNum)
```


whether this is worth it or not, I don't know.

**Comment by hashbackup** [comment]
Changing assert to doAssert is definitely better than hanging.  I didn't know the difference, other than sometimes I see one, other times the other.

IMO, it only makes sense to pass the expected number of items because hash table size depends on implementation details that the user doesn't know and may change in the future - mainly the load factor that triggers a resize.  For Robin Hood hashing, the load factor can easily go up to 90%.  For HashBackup's dedup table, I run it at 80% with good results, though with a completely different setup.  In the example, rightSize determines that Nim's hash table needs 32 entries to be able to insert 16 elements without a resize.  Checking whether initialSize is a power of 2 and forcing it if not doesn't make sense, because while it won't hang, it will resize when adding 16 items, defeating the purpose of specifying the size.

Also, making an implementation detail of "must be a power of 2", a documented requirement as it is now, could lead to future compatibilities problems if a new implementation requires a power of 3, a prime number, etc.
 
What initTable really needs to know is "did initialSize come from rightSize()?"  My suggested fix is designed to detect that - not whether the size is a power of 2.

I added a couple more examples to the test program.  Here are the results for passing 16, 31, and 32 to init, by any of the 3 ways (int, nextPowerOfTwo, and rightSize):
```
tableSize is 32
tableSize is 32
tableSize is 32

tableSize is 64
tableSize is 64
tableSize is 64

tableSize is 64
tableSize is 64
tableSize is 64
```
I don't think it's worth it to add a new interface so size and # elements can be specified, since I don't think users should ever specify size directly.

**Comment by timotheecour** [comment]
> I don't think it's worth it to add a new interface so size and # elements can be specified, since I don't think users should ever specify size directly.

again, the problem is backward compatibility for all the code out there; if `proc initTable*[A, B](initialSize = defaultInitialSize): Table[A, B] =` changes the meaning of `initialSize` from table size to expected number of elements, it will cause over-allocations eg:

```
let a = initTable[A,B](2 shl 10)
# before: allocates 2 shl 10, which is what application expects
# after: allocates rightSize(2 shl 10) = 2 shl 11, which may allocate too much memory for that application, maybe even running out of memory in some cases
```

which is why if such a change is done, it needs another API (which can mean an extra optional param in existing API)

**Comment by hashbackup** [comment]
Yeah, that is a possibility.  I'd argue that anyone using initTable with an initialSize other than coming from rightSize, doesn't really know what they're getting.  If initialSize was specified in MB, it would make sense.  But when specifying the number of slots in the hash table, the way it is now, they'd have to know how the hash table is implemented to know how much memory theirs will use.

Or, another possibility is that someone knew how many elements they wanted in the table, looked up the definition of rightSize, ran those calculations, rounded up to a power of 2, and used that.  Then I they wouldn't like it if it got doubled on them in a new release.

Instead of adding another init API, I'd prefer it get left alone and just use doAssert to bomb.  I think it's dumb as is, but a new API makes it even worse.

**Comment by c-blake** [comment]
First and foremost, it bears mentioning that the main speed benefit of `rightSize` is to avoid successive rehashes growing the table up.  The penalty of always applying it and making the table 2x larger for po2 callers (or already calling-`rightSize`-callers) is just to spend 2x the RAM which could even make things faster if that RAM is available/cachable/things are not right at a cliff in the memory latency hierarchy.  Were someone to change that threshold from 2/3 to 1/2 or 3/4 or growth amount to 4x from 2x (both opaque to client code), we would face similar questions.  Often in documentation one sees language like "size hint" to not get too trapped by such implementation details.  Even resizing based upon load factor could change to instead being being triggered by search depth, as I've been experimenting with in https://github.com/c-blake/adix.  Neither growth trigger is difficult, but they have different properties.  Search depth makes max search time more predictable but `rightSize` a total guess.  I get near 100% utilization on Robin Hood hashing.  Meanwhile occupancy makes `rightSize` predictable, but search times a guess.

I originally left things the fragile way out of a perhaps misguided sense of "compatibility" when I added `rightSize`.  I mostly thought that API/capability *at all* (given opaqueness and occupancy triggered resize) was more important than worrying about how it got engaged or a big discussion.

There is also this regrettable need to qualify `sets.rightSize` vs. `tables.rightSize` even though they do the same thing (induced by `sets` and `tables` being different modules).

Anyhow, what we should probably do now is just define exported `rightSize*` to be the ***identity function*** and define and always apply a **new, unexported** `rightSizeInternal` to do the actual calculation.  We could also **deprecate** `rightSize*` and maybe remove it in Nim-2.0.  Though code doing its own po2 was always implicitly assuming 2/3 (not 1/2 or 3/4 etc.), even such "performance risky" code can be supported with a `when defined(noAutoRightSize):` guard around the call to `rightSizeInternal`.  And we should clarify that `initialSize` is only a "size hint" to cover future commitments/expectations in case we want to guess about size instead of max search depth.  And maybe update any Nim-internal call sites of those init functions, of course.

Correct me if I'm wrong, but I think this covers all objections.  It only creates a (maybe faster, maybe slower) size situation for code that A) already ignores `rightSize` and does its own po2 calc ***and*** B) has a maintainer who does not read Changelog to see to add `d:noAutoRightSize` or otherwise adapt their code.  Everyone else gets a simpler life in three ways - A) no symbol qualification or even a need to invoke the symbol, B) the size hint being about the only size they should really need to know, and C) the API being more like other prog.lang runtimes.

EDIT: not sure if we can deprecate the symbols used in such `when` guards, but that could also be a path for compiler assisted code evolution later on.

**Comment by Araq** [comment]
> There is also this regrettable need to qualify sets.rightSize vs. tables.rightSize even though they do the same thing (induced by sets and tables being different modules).

That one is solvable, move ``rigthSize`` to its own module and have sets and tables re-export it then no ambiguities arise.

**Comment by c-blake** [comment]
Well, sure..you might have me on part 1 of A) :-)  But no symbol at all is even simpler.  

Anyway, that module improvement is independent and could be done anyway, though for a deprecated symbol pointing to an inline identity function a whole module might be of questionable value. ;-)

**Comment by c-blake** [comment]
We probably don't *really* need to update any pre-sizing po2 inits unless we care about preserving such guesses exactly which may also preserve hash order **if** no resize happens, a fragile situation unlikely to be relied upon -- except in tests.

For the few cases that might matter just dividing the explicit po2 guess restores the exact guess which is an easier patch than other times we've changed hash orders.  Code using `rightSize` will work exactly as before.  You don't really *need* to deprecate the do-nothing call.

`-d:noAutoRightSize` is just a transition path/to appease perfect compatibility.

I think it's a way forward with only a couple small updates in some test code here & there with a less implementation-tied and user-friendly result.

**Comment by hashbackup** [comment]
Fixed by #14926, thanks!
